src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -5394,6 +5417,253 @@ def delete_deployment(self, deployment_id: UUID) -> None:
+                CuratedVisualizationSchema.visualization_index
+                == visualization_index
+            )
+            .where(CuratedVisualizationSchema.resource_id == resource_id)
+            .where(
+                CuratedVisualizationSchema.resource_type == resource_type.value
+            )
+        ).first()
+        if existing is not None:
+            raise EntityExistsError(
+                "A curated visualization for this resource already exists "
+                "for the specified artifact version and visualization index."
+            )
+
+    def create_curated_visualization(
+        self, visualization: CuratedVisualizationRequest
+    ) -> CuratedVisualizationResponse:
+        """Persist a curated visualization link.
+
+        Args:
+            visualization: The curated visualization to persist.
+
+        Returns:
+            The persisted curated visualization.
+
+        Raises:
+            IllegalOperationError: If the resource does not belong to the same project as the curated visualization.
+            KeyError: If the resource does not exist.
+        """
+        with Session(self.engine) as session:
+            self._set_request_user_id(
+                request_model=visualization, session=session
+            )
+
+            artifact_version: ArtifactVersionSchema = self._get_schema_by_id(
+                resource_id=visualization.artifact_version_id,
+                schema_class=ArtifactVersionSchema,
+                session=session,
+            )
+
+            # Validate explicitly provided project before defaulting
+            if visualization.project:
+                if visualization.project != artifact_version.project_id:
+                    raise IllegalOperationError(
+                        "Curated visualizations must target the same project as "
+                        "the artifact version."
+                    )
+                project_id = visualization.project
+            else:
+                project_id = artifact_version.project_id
+
+            self._validate_curated_visualization_index(
+                session=session,
+                artifact_version_id=visualization.artifact_version_id,
+                visualization_index=visualization.visualization_index,
+            )
+
+            resource_schema_map: Dict[
+                VisualizationResourceTypes, Type[BaseSchema]
+            ] = {
+                VisualizationResourceTypes.DEPLOYMENT: DeploymentSchema,
+                VisualizationResourceTypes.MODEL: ModelSchema,
+                VisualizationResourceTypes.PIPELINE: PipelineSchema,
+                VisualizationResourceTypes.PIPELINE_RUN: PipelineRunSchema,
+                VisualizationResourceTypes.PIPELINE_SNAPSHOT: PipelineSnapshotSchema,
+                VisualizationResourceTypes.PROJECT: ProjectSchema,
+            }
+
+            if visualization.resource_type not in resource_schema_map:
+                raise IllegalOperationError(
```

Comment:
I think `ValueError` might be better, it's what we usually raise and results in a `422` response which I think is appropriate here with an invalid request that we can't process?